src/Stratis.Features.SQLiteWalletRepository/External/ITransactionsToLists.cs

```diff
@@ -124,50 +124,59 @@ public bool ProcessTransactions(IEnumerable<Transaction> transactions, HashHeigh
 
                     var destinations = this.GetDestinations(txOut.ScriptPubKey);
 
-                    bool isChange = destinations.Any(d => addressesOfInterest.Contains(d, out AddressIdentifier address2) && address2.AddressType == 1);
+                    bool isChange = destinations.Any(d => addressesOfInterest.Contains(d.ScriptPubKey, out AddressIdentifier address2) && address2.AddressType == 1);
 
                     if (addSpendTx)
                     {
                         // TODO: Why is this done? If the receipt is not to one of our addresses (i.e. identified in the loop coming next) then why bother trying to record it?
                         this.RecordReceipt(block, null, txOut, tx.IsCoinBase | tx.IsCoinStake, blockTime ?? currentTime, txId, i, isChange);
                     }
 
-                    foreach (Script pubKeyScript in destinations)
+                    foreach (TxDestination destination in destinations)
                     {
+                        var pubKeyScript = destination.ScriptPubKey;
                         bool containsAddress = addressesOfInterest.Contains(pubKeyScript, out AddressIdentifier address);
 
                         // Paying to one of our addresses?
-                        if (containsAddress)
+                        if (containsAddress && address != null)
                         {
+                            // This feature, by design, is agnostic of the type of template being processed.
+                            // This type of check is good to have for cold staking though but is catered for in broader terms.
+                            // I.e. don't allow any funny business with keys being used with accounts they were not intended for.
+                            if (destination is IAccountRestrictedKeyId accountRestrictedKey)
```

Comment:
If `destination` is of type `TxDestination` I can't see how this can ever be true?

Where does `TxDestination` implement  `IAccountRestrictedKeyId`?

Reply:
Look in the bottom of the first code snippet.